tmail-backend/mailbox/opensearch/src/main/java/com/linagora/tmail/mailbox/opensearch/TmailCriterionConverter.java

```diff
@@ -0,0 +1,624 @@
+        registerHeaderOperatorConverters();
+    }
+
+    @VisibleForTesting
+    public TmailCriterionConverter() {
+        this.criterionConverterMap = new HashMap<>();
+        this.headerOperatorConverterMap = new HashMap<>();
+        this.textFuzzinessSearchValue = evaluateFuzzinessValue(DEFAULT_TEXT_FUZZINESS_SEARCH);
+        this.useQueryStringQuery = false;
+
+        registerCriterionConverters();
+        registerHeaderOperatorConverters();
+    }
+
+    private String evaluateFuzzinessValue(boolean textFuzzinessSearchEnable) {
+        if (textFuzzinessSearchEnable) {
+            return "AUTO";
+        }
+        return "0";
+    }
+
+    private void registerCriterionConverters() {
+        registerCriterionConverter(SearchQuery.FlagCriterion.class, this::convertFlag);
+        registerCriterionConverter(SearchQuery.UidCriterion.class, this::convertUid);
+        registerCriterionConverter(SearchQuery.MessageIdCriterion.class, this::convertMessageId);
+        registerCriterionConverter(SearchQuery.ConjunctionCriterion.class, this::convertConjunction);
+        registerCriterionConverter(SearchQuery.HeaderCriterion.class, this::convertHeader);
+        registerCriterionConverter(SearchQuery.SubjectCriterion.class, this::convertSubject);
+        registerCriterionConverter(SearchQuery.TextCriterion.class, this::convertTextCriterion);
+        registerCriterionConverter(SearchQuery.CustomFlagCriterion.class, this::convertCustomFlagCriterion);
+
+        registerCriterionConverter(SearchQuery.AllCriterion.class,
+            criterion -> new MatchAllQuery.Builder().build().toQuery());
+
+        registerCriterionConverter(SearchQuery.ModSeqCriterion.class,
+            criterion -> createNumericFilter(JsonMessageConstants.MODSEQ, criterion.getOperator()));
+
+        registerCriterionConverter(SearchQuery.SizeCriterion.class,
+            criterion -> createNumericFilter(JsonMessageConstants.SIZE, criterion.getOperator()));
+
+        registerCriterionConverter(SearchQuery.InternalDateCriterion.class,
+            criterion -> dateRangeFilter(JsonMessageConstants.DATE, criterion.getOperator()));
+
+        registerCriterionConverter(SearchQuery.SaveDateCriterion.class,
+            criterion -> dateRangeFilter(JsonMessageConstants.SAVE_DATE, criterion.getOperator()));
+
+        registerCriterionConverter(SearchQuery.AttachmentCriterion.class, this::convertAttachmentCriterion);
+        registerCriterionConverter(SearchQuery.MimeMessageIDCriterion.class, this::convertMimeMessageIDCriterion);
+        registerCriterionConverter(SearchQuery.ThreadIdCriterion.class, this::convertThreadIdCriterion);
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T extends SearchQuery.Criterion> void registerCriterionConverter(Class<T> type, Function<T, Query> f) {
+        criterionConverterMap.put(type, (Function<SearchQuery.Criterion, Query>) f);
+    }
+
+    private void registerHeaderOperatorConverters() {
```

Comment:
Stupid idea: it would be wonderful to be able to reuse the subparts of criterion converter without reimplementing it all.

How about changing visibility of everything used in `registerCriterionConverters` in `protected` so that wedo not need to duplicate things here and automatically benefits from James enhancements ?

Reply:
Actually after working on it... All the code logic in defined in DefaultCriterionConverter... but we implement CriterionConverter which is an interface?

Except if you want to extend DefaultCriterionConverter instead? But honestly it defeats the logic of the interface